api/envoy/config/listener/v3/listener_components.proto

```diff
@@ -193,6 +318,46 @@ message FilterChainMatch {
   //   and matching on values other than ``h2`` is going to lead to a lot of false negatives,
   //   unless all connecting clients are known to use ALPN.
   repeated string application_protocols = 10;
+
+  // A list of filter chain matching predicate extensions specifying the order
+  // of selection for the most specific filter chain. The filter chains in a
+  // single listener must have the identical order of predicate extension type
+  // URLs. For backwards compatibility, if the existing fields in the filter
+  // chain match are specified, the fields are converted to their corresponding
+  // filter chain matching predicates in the following default matching order:
+  // 1. :ref:`DestinationPort <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DestinationPort>`.
+  // 2. :ref:`DestinationIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DestinationIP>`.
+  // 3. :ref:`SNI <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SNI>`.
+  // 4. :ref:`TransportProtocol <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.TransportProtocol>`.
+  // 5. :ref:`ApplicationProtocol <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.ApplicationProtocol>`.
+  // 6. :ref:`DirectSourceIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.DirectSourceIP>`.
+  // 7. :ref:`SourceType <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourceType>`.
+  // 8. :ref:`SourceIP <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourceIP>`.
+  // 9. :ref:`SourcePort <envoy_v3_api_msg_config.listener.v3.FilterChainMatch.SourcePort>`.
+  //
+  // At each step, the predicate is evaluated for each filter chain. If the
+  // predicate returns false, the filter chain is eliminated from subsequent
+  // consideration. If the predicate returns true, then the decision is based
+  // on the ``fallthrough`` flag. If there is another more specific match for
+  // the same predicate and the flag is not set, then the less specific match
+  // is eliminated. If the flag is set, then it remains as a candidate for the
+  // next step.
+  //
+  // For example, consider SNI ``www.example.com``and two filter chains with
+  // the predicates ```*.example.com``` and ```*.com```. If ``fallthrough``
+  // flag is not set then only the filter chain ```*.example.com``` matches. If
+  // ``fallthrough`` flag is set on ```*.com```, then both filter chains match.
+  // In general, the order of specificity is domain specific as defined by the
+  // predicate extension. The flag should be set when matching on multiple
+  // properties in order for a default chain to apply without explicit specific
+  // matching of the first properties in the list.
```

Comment:
Can you beef this up even a bit more with a worked example? I'm having a hard time wrapping my head around the case in which someone would want multiple matches, and also, if there are multiple matches, what does that mean for trie/sequential based matching? Does it basically mean that if there are multiple matches we keep recursing downward on N tree branches to see if there are further reductions in the search space?

Reply:
Related to trie/sequential-based matching, would it make sense for this to use the new generic matching API that @snowp has been working on, rather than inventing yet another matcher structure?

https://github.com/cncf/xds/blob/main/xds/type/matcher/v3/matcher.proto

Reply:
@mattklein123 Adding more explanations. The extra matches are already inevitable in the existing API because of overlapping wildcards. What we're trying to change is making it more succinct for the control plane to express.  I think there is normally no need to have multiple matches at the very end, but in the middle of matching, the "default" case has to be kept around in the search space until the full "special" case is matched. What happens now, is that control plane has to provide many identical "default" cases for each step of "special" case matching, and that quickly proliferates. 

I think we can implement the fallthrough wildcard internally with trie upward propagation. E.g. if there is a chain FC1 wildcard "*.com" and trie node "example.com" for FC2, then we automatically add FC1 to node "example.com". This what control plane would have to do explicitly.

@markdroth I took a look at the generic matcher API. I think it's rather difficult to use. Our main issue with the existing state is that it's too hard to use right (e.g. Istio will not migrate to the status quo API and will stay on the "undeprecated" field). Make it more abstract does not seem to help with usability IMHO. It also seems L7-oriented right now, and the set of matchers is rather distinct. I can be convinced the other way, if we can construct some examples that are easy to comprehend.

Reply:
I think the semantics of the generic matching API are actually much clearer and easier to understand, because instead of memorizing a set of precedence rules, you can explicitly encode the precedence in the matcher structure.  It's actually not abstract at all; it's very explicitly and precise about what it represents.

The generic matching API was specifically designed to be extensible, so that you can plug in whatever inputs and whatever new match types you need.  It was intentionally designed *not* to be L7-specific.  For background, see https://docs.google.com/document/d/1G4g-6q0IArz_ERgqixzZCM0-wbexFYuBUvT9hkV7QRU/edit.

I really don't think we should be reinventing this wheel yet again.  In the long run, we want all matching in xDS to move to the new generic matching API.

Reply:
It hadn't occurred to me to use the generic matching API for this, but given that it inherently supports sub-linear matching, I tend to agree with @markdroth that we should see if we can use it. Can you mock up what that might look like and we can discuss? If there are usability issues with that API we should fix them.

Reply:
@mattklein123 I am not sure where to begin because the matcher API does not specify the "wildcard" matching behavior with the more specific candidate winning. That seems to be the crux of the problem in this context.

Reply:
@snowp any thoughts on ^?

Reply:
While that doesn't exist today I would expect us to be able to implement virtually any matching semantics using a matcher tree `custom_match`, though that would have to be implemented (not hard, just wiring up some extension lookups). We could also consider making it a built in matcher alongside exact_match/custom_match if we think it deserves that

Reply:
Yeah, plugging that in via the `custom_match` field probably makes the most sense for now, since it seems fairly specific to Listener matching.  We could promote it to a first-class matcher type later if we find that it's also useful in other contexts (like maybe VirtualHost matching? not sure).

Reply:
Can someone explain how to change matching semantics in tree matcher? The problem is that we don't want eager matching in listener filter chain matching fallthrough case. Projecting to the HTTP semantics, consider the two routes:
1) Path: /path exact, Header X = Y
2) Path: / prefix

We want route 2) to apply to route /path with header X = Z. So we don't want an eager pruning step at the path node. We want to match on the combination of path and header.

I also don't think that using "custom" matching for a first class object in Envoy is appropriate. If every filter chain match condition has to use "custom", that adds quite a bit of verbosity and hurts feature use.

Reply:
Also, I understand that any trie decision tree could be encoded in abstract sense. But I'm not seeing the succinct representation that avoids repeated (2^n) filter chains with slightly varied matching conditions.

Reply:
Note that use of an API extension for an input or a matcher does not imply that the functionality is not a first-class citizen; it simply implies that it's a protocol-specific input or match type that is not built into the generic matching framework itself.  For example, HTTP header inputs are an extension, and that is very much a first-class citizen in the API.

The idea of the generic matching API is that you construct a tree of matchers to represent things like AND and OR operations, where each individual node in the tree can be either a `MatcherList` (for linear matching) or a `MatcherTree` (for sublinear matching).  In either case, if a match is found, we use the corresponding `OnMatch`, which can be either a protocol-specific action or a nested matcher (i.e., another node in the tree of matchers).  If no match is found in a node, there is another `OnMatch` called `on_no_match` that will be used if populated; if it's not populated, then the node is considered not to match, and matching resumes from there.

For the example you cited, you'd structure it as the top node being a `MatcherList`, where the matchers in the list are:

1. Match on path=/path.  If that matches, use a nested matcher that checks for the header X = Y.  (If the nested matcher does not match and does not have an `on_no_match` field, then we will move on to the next entry in the list.)
2. Match on path prefix /.

Alternatively, if you wanted that second rule to be applied to anything that was not matched previously (i.e., you wanted it to apply to all requests, not just those with prefix="/"), you could move it out of the list and put it in the `on_no_match` field in the top node.

Reply:
Is there a way to build a matcher tree internally in Envoy? I think we have to use an outer MatcherList because of the defaulting behavior. But then each item might have overlapping predicates. For example,

1. MatchList item1: Match on destination port 80.
2. MatchList item2-1000: Match on other ports in the range of 1-1000.
1001. MatchList item: Match on destination port range 80-1000.

The linear semantics is sub-optimal here at the outer level. We want an implicit lookup tree, but we also don't want directly go into individual node 1-1000 for the default case 1001.

Reply:
> Is there a way to build a matcher tree internally in Envoy

You should be able to build up any config internally that you like, though if there are API issues with the type of matching you want to do we should sort that out.

Reply:
I don't know why we'd want to build the tree internally.  I think the goal here should be to allow the API to explicitly configure how the matching should be done, so that there's no more counter-intuitive magic here.

I can tell you from when we implemented this logic in gRPC recently that the current matching behavior here is very difficult to understand.  The behavior is that it basically hard-codes the order of matching to the following list:
1. destination port
2. destination IP
3. server name
4. transport protocol
5. application protocol
6. connection source type
7. source IP
8. source port

Trying to reason through how a given connection will be matched to a filter chain requires that you memorize that list, and even then, it takes real thought and it's very confusing to implement correctly.  I think making this explicit in the API would be a big improvement.

For the destination port example, I think it should be fairly simply to define a new type of sublinear matcher that is keyed on port ranges.  You could have one entry for ports 1-79, one for port 80, and one for ports 81-10000.  Any given incoming connection would fit into exactly one of those three categories, so there's no need for linear matching behavior.

The only complication I see here is that there will wind up being multiple leaves in the tree for each filter chain, and we don't want to have to duplicate the filter chain for each leaf that uses it.  But I think this can be addressed fairly simply by moving the actual filter chain definitions to a separate map, keyed by some opaque name, and then having the matcher leaves refer to the keys in that map.  That way, whenever multiple leaves refer to the same chain, the only duplication is the opaque name.